plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+	HPI_EVENT_HARD_RESET_SENT,
+	HPI_EVENT_SOFT_RESET_SENT,
+	HPI_EVENT_CABLE_RESET_SENT,
+	HPI_EVENT_SOURCE_DISABLED,
+	HPI_EVENT_SENDER_TIMEOUT,
+	HPI_EVENT_VDM_NO_RESPONSE,
+	HPI_EVENT_UNEXPECTED_VOLTAGE,
+	HPI_EVENT_ERROR_RECOVERY,
+	HPI_EVENT_EMCA_DETECT			= 0xA6,
+	HPI_EVENT_RP_CHANGE_DETECT		= 0xAA,
+	HPI_EVENT_TB_ENTERED			= 0xB0,
+	HPI_EVENT_TB_EXITED
+}HPIResp;
+
+typedef enum
+{
+	/* register region */
+	HPI_REG_PART_REG			= 0,
+	/* data memory for device section */
+	HPI_REG_PART_DATA			= 1,
+	/* flash memory	*/
+	HPI_REG_PART_FLASH			= 2,
+	/* read data memory for port section */
+	HPI_REG_PART_PDDATA_READ		= 4,
+	/* write data memory for port section */
+	HPI_REG_PART_PDDATA_WRITE		= 8
+} HPIRegPart;
+
+typedef struct __attribute__((packed))
+{
+	guint16	event_code;
+	guint16	event_length;
+	guint8	event_data[128];
+}HPIEvent;
+
+static guint16
+fu_ccgx_hpi_reg_addr_gen (guint8 section, guint8 part, guint8 reg_addr)
+{
+	return (guint16) ( (section << 12) | (part << 8) | reg_addr);
+}
+
+static gboolean
+fu_ccgx_hpi_reg_read (FuDevice *device,
+		      CyI2CDeviceHandle *i2c_handle,
+		      guint8 hpi_addr_byte,
+		      guint16 reg_addr,
+		      guint8 *reg_data,
+		      guint16 size,
+		      GError **error)
+{
+	guint32 i;
+	g_autofree guint8 *write_buffer = NULL;
+
+	CyDataBuffer data_buffer = {0};
+	CyI2CDataConfig data_cfg = {0};
+
+	g_return_val_if_fail (i2c_handle != NULL, FALSE);
+	g_return_val_if_fail (reg_data != NULL, FALSE);
+	g_return_val_if_fail (hpi_addr_byte > 0, FALSE);
+
+	write_buffer = g_malloc0 (hpi_addr_byte + 1);
+
+	g_return_val_if_fail (write_buffer != NULL, FALSE);
+
+	data_cfg.is_stop_bit = FALSE;
+	data_cfg.is_nak_bit = TRUE;
+	data_buffer.length = hpi_addr_byte;
+
+	for (i = 0; i < hpi_addr_byte; i++) {
+		write_buffer[i] = (guint8)(reg_addr >> (8*i));
+	}
+
+	data_buffer.buffer = write_buffer;
+	data_buffer.transfer_count = 0;
+
+	if (!fu_ccgx_i2c_write(device, i2c_handle, &data_cfg, &data_buffer, error)) {
+		g_prefix_error (error, "fu_ccgx_hpi_reg_read - write error: ");
+		return FALSE;
+	}
+
+	data_cfg.is_stop_bit = TRUE;
+	data_cfg.is_nak_bit = TRUE;
+
+	data_buffer.buffer = reg_data;
+	data_buffer.length = size;
+
+	if (!fu_ccgx_i2c_read(device, i2c_handle, &data_cfg, &data_buffer, error)) {
+		g_prefix_error (error, "fu_ccgx_hpi_reg_read - read	error: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_hpi_reg_write (FuDevice *device,
+		       CyI2CDeviceHandle *i2c_handle,
+		       guint8 hpi_addr_byte,
+		       guint16 reg_addr,
+		       guint8 *reg_data,
+		       guint16 size,
+		       GError **error)
+{
+	g_autofree guint8 *write_buffer = NULL;
+
+	CyDataBuffer data_buffer;
+	CyI2CDataConfig data_cfg;
+
+	g_return_val_if_fail (i2c_handle != NULL, FALSE);
+	g_return_val_if_fail (reg_data != NULL, FALSE);
+	g_return_val_if_fail (hpi_addr_byte > 0, FALSE);
+
+	write_buffer = g_malloc0 (size + hpi_addr_byte + 1);
+
+	g_return_val_if_fail (write_buffer != NULL, FALSE);
```

Comment:
I think this specific `g_return_val_if_fail` is probably overkill.

Reply:
Is it right for me to remove g_return_val_if_fail and not to check whether write_buffer is NULL?